File: backend/generate_build_info.py
#!/usr/bin/env python3
"""
Generate build info for the backend
"""
import json
import os
import subprocess
from datetime import datetime, timezone
from pathlib import Path


def get_git_commit() -> str:
    """Get current git commit hash"""
    try:
        result = subprocess.run(
            ["git", "rev-parse", "HEAD"],
            capture_output=True,
            text=True,
            check=True
        )
        return result.stdout.strip()
    except (subprocess.CalledProcessError, FileNotFoundError):
        return "unknown"


def get_git_branch() -> str:
    """Get current git branch"""
    try:
        result = subprocess.run(
            ["git", "rev-parse", "--abbrev-ref", "HEAD"],
            capture_output=True,
            text=True,
            check=True
        )
        return result.stdout.strip()
    except (subprocess.CalledProcessError, FileNotFoundError):
        return "unknown"


def generate_build_info() -> dict:
    """Generate build information"""
    return {
        "version": os.getenv("BUILD_VERSION", "dev"),
        "buildNumber": os.getenv("BUILD_NUMBER", "local"),
        "gitCommit": os.getenv("GIT_COMMIT", get_git_commit()),
        "gitBranch": get_git_branch(),
        "environment": os.getenv("ENVIRONMENT", "development"),
        "buildTime": datetime.now(timezone.utc).isoformat(),
        "service": "backend"
    }


def main() -> None:
    """Main function"""
    build_info = generate_build_info()

    # Write to build-info.json
    build_info_path = Path(__file__).parent / "build-info.json"
    with open(build_info_path, "w") as f:
        json.dump(build_info, f, indent=2)

    print(f"✅ Build info generated: {build_info_path}")
    print(json.dumps(build_info, indent=2))


if __name__ == "__main__":
    main()

File: frontend/generate-build-info.js
#!/usr/bin/env node
/**
 * Generate build info for the frontend
 */
import fs from 'fs';
import path from 'path';
import { execSync } from 'child_process';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

function getGitCommit() {
  try {
    return execSync('git rev-parse HEAD', { encoding: 'utf8' }).trim();
  } catch (error) {
    return 'unknown';
  }
}

function getGitBranch() {
  try {
    return execSync('git rev-parse --abbrev-ref HEAD', { encoding: 'utf8' }).trim();
  } catch (error) {
    return 'unknown';
  }
}

function generateBuildInfo() {
  return {
    version: process.env.BUILD_VERSION || 'dev',
    buildNumber: process.env.BUILD_NUMBER || 'local',
    gitCommit: process.env.GIT_COMMIT || getGitCommit(),
    gitBranch: getGitBranch(),
    environment: process.env.ENVIRONMENT || process.env.NODE_ENV || 'development',
    buildTime: new Date().toISOString(),
    service: 'frontend'
  };
}

function main() {
  const buildInfo = generateBuildInfo();

  // Ensure public directory exists
  const publicDir = path.join(__dirname, 'public');
  if (!fs.existsSync(publicDir)) {
    fs.mkdirSync(publicDir, { recursive: true });
  }

  // Write to build-info.json
  const buildInfoPath = path.join(publicDir, 'build-info.json');
  fs.writeFileSync(buildInfoPath, JSON.stringify(buildInfo, null, 2));

  console.log(`✅ Build info generated: ${buildInfoPath}`);
  console.log(JSON.stringify(buildInfo, null, 2));
}

// Run main function if this is the main module
if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}

export { generateBuildInfo };